### DISCUSSION TITLE
How to compile OpenCV with CUDA support?

### DISCUSSION THREAD

**Discussion by EmilPi** [apache/incubator-mxnet#20553] [Q&A] [answered]
When trying to use `mxnet.gluon.data.vision.transforms` I get error
`MXNetError: Operator _cvimresize is not implemented for GPU.`.

As a workaround, I am sending data to CPU, augmenting it and then sending back to GPU. This is horribly slow.
Do I need to compile OpenCV separately or should I modify config.cmake, or both? Is there any guide on this? What I found is very old.

**Comment by szha** [comment, marked as answer]
@EmilPi it would require integrating the opencv GPU support in mxnet image API to use it natively with GPU arrays. For now you can consider using the OpenCV python API with GPU support for data augmentation, and then use [from_dlpack](https://mxnet.apache.org/versions/1.8.0/api/python/docs/api/ndarray/ndarray.html#mxnet.ndarray.from_dlpack) and [to_dlpack_for_write](https://mxnet.apache.org/versions/1.8.0/api/python/docs/api/ndarray/ndarray.html#mxnet.ndarray.to_dlpack_for_write) for zero-copy loading and sharing of ndarray.

**Reply by EmilPi** [reply]
Ok, I will try. 谢谢！

**Reply by EmilPi** [reply]
Unluckily, I could not use OpenCV for that, but that means, I have to use other augmentation library.

**Comment by EmilPi** [comment]
Is there any example of sharing mxnet <-> OpenCV GpuMat data? What am I doing wrong?

I am trying to
1. get data
2. pass it to opencv GpuMat and augment
3. get MXNet GPU array from opencv using mxnet.ndarray.from_dlpack(gpu_mat)

But this code

    import cv2
    import mxnet as mx
    import mxnet.ndarray as nd
    
    ctx = mx.gpu()
    
    w, h, d = 8, 8, 3
    angle, scale = 45, .9
    M = cv2.getRotationMatrix2D((w / 2, h / 2), angle, scale)
    image = nd.random.uniform(shape=(w, h, d)).as_in_context(ctx)
    
    image_ = image.asnumpy()  # Let's say this is loaded from a disk
    cv_gpumat = cv2.cuda_GpuMat(image_)  # CPU -> GPU
    cv_gpumat_augmented = cv2.cuda.warpAffine(cv_gpumat, M, (w, h))  # GPU -> GPU
    # next line fails :(
    nd_arr = nd.from_dlpack(cv_gpumat_augmented)  # GPU -> GPU

Fails with `AssertionError: Invalid DLPack Tensor. DLTensor capsules can be consumed only once.` error.

I also tried

    image_cv = nd.to_dlpack_for_write(image)
    cv_gpumat = cv2.cuda_GpuMat()
    cv_gpumat.upload(image_cv)
    image_cv_augmented = cv2.cuda.warpAffine(cv_gpumat, M, (w, h))
    image_augmented = nd.from_dlpack(image_cv_augmented)

but this fails with

    cv2.error: OpenCV(4.5.3-dev) :-1: error: (-5:Bad argument) in function 'upload'
    > Overload resolution failed:
    >  - arr is not a numpy array, neither a scalar
    >  - Expected Ptr<cv::cuda::GpuMat> for argument 'arr'
    >  - Expected Ptr<cv::UMat> for argument 'arr'
